internal/pkg/earpointer/analysis.go

```diff
@@ -0,0 +1,649 @@
+	// *address".
+	// The address can be a (memory) address, a global variable, or a free
+	// variable.
+	if mayShareObject(store.Val) {
+		vis.processAddressToValue(store.Addr, store.Val)
+	}
+}
+
+func (vis *visitor) visitPhi(phi *ssa.Phi) {
+	if !typeMayShareObject(phi.Type()) {
+		return
+	}
+	for _, e := range phi.Edges {
+		vis.unifyLocals(phi, e)
+	}
+}
+
+func (vis *visitor) visitMapUpdate(update *ssa.MapUpdate) {
+	vis.processMemoryIndexAccess(update.Value, update.Map, update.Key)
+}
+
+func (vis *visitor) visitConvert(convert *ssa.Convert) {
+	vis.unifyLocals(convert, convert.X)
+}
+
+func (vis *visitor) visitChangeInterface(change *ssa.ChangeInterface) {
+	vis.unifyLocals(change, change.X)
+}
+
+func (vis *visitor) visitChangeType(change *ssa.ChangeType) {
+	vis.unifyLocals(change, change.X)
+}
+
+func (vis *visitor) visitUnOp(op *ssa.UnOp) {
+	switch op.Op {
+	case token.ARROW: // channel read
+		// Use array read to simulate channel read.
+		vis.processHeapAccess(op, op.X, anyIndexField)
+	case token.MUL:
+		// v = *addr
+		// The LHS is a register; the RHS can be a (memory) address, a global
+		// variable, or a free variable.
+		// t1 = *addr is implemented as "addr[directPointToField] = t1",
+		// i.e., address addr points to value t1.
+		vis.processAddressToValue(op.X, op)
+	}
+}
+
+func (vis *visitor) visitBinOp(op *ssa.BinOp) {
+	// Numeric binop will be skipped.
+	// The only binop that matters now is "+" that concatenates strings.
+	if op.Op == token.ADD {
+		vis.unifyLocals(op, op.X)
+		vis.unifyLocals(op, op.Y)
+	}
+}
+
+func (vis *visitor) visitTypeAssert(assert *ssa.TypeAssert) {
+	if !assert.CommaOk {
+		// t1 = typeassert t0.(*int)
+		vis.unifyLocals(assert, assert.X)
+	}
+	// The comma case will be processed by the subsequent extract instruction.
+}
+
+func (vis *visitor) visitExtract(extract *ssa.Extract) {
+	switch base := extract.Tuple.(type) {
+	case *ssa.TypeAssert:
+		// t1 = typeassert,ok t0.(*int)
+		// t2 = extract t1 0
+		if base.CommaOk && extract.Index == 0 {
+			vis.unifyLocals(extract, base.X)
+		}
+	case *ssa.Next:
+		// t1 = range t0
+		// t2 = next t1   // return (ok bool, k Key, v Value)
+		// t3 = extract t2 ?
+		if rng, ok := base.Iter.(*ssa.Range); ok {
+			if mayShareObject(extract) && extract.Index == 2 { // extract the value v
+				// The value from the map t0 flows into the field: t0[*] = v.
+				vis.processHeapAccess(extract, rng.X, anyIndexField)
+			}
+		}
+	case *ssa.Lookup:
+		// t2 = t0[t1],ok
+		// t3 = extract t2 ?
+		if base.CommaOk && extract.Index == 0 {
+			vis.processMemoryIndexAccess(extract, base.X, base.Index)
+		}
+	default: // Other cases: model "r1 = extract r0 #n" as "r0["n"] = r1".
+		field := Field{Name: strconv.Itoa(extract.Index)}
+		vis.processHeapAccess(extract, base, field)
+	}
+}
+
+func (vis *visitor) visitNext(next *ssa.Next) {
+	// The next instruction will be processed by the subsequent extract
+	// instruction.
+}
```

Comment:
The case that would use this has the comment explaining that it gets processed by the `extract`, leaving this method unused.